### PR TITLE
MKL: Provide scalapack

### PR DESCRIPTION
**Pull request by Zzzoom**
This PR turns intel-mkl and intel-parallel-studio into scalapack providers...for most MPIs, anyway.

A couple of things are iffy:

- scalapack should be provided when ^mpi (is this valid?). There's a test for that later that raises an exception if there are no known MPI implementations in the spec, though.
- It doesn't support Intel MPI because afaik there's no way to distinguish between ^intel-parallel-studio being pulled for Intel MPI or the rest of the stuff in the package. (including scalapack itself)

### CODE REVIEW

File: var/spack/repos/builtin/packages/intel-mkl/package.py [hunk 1]
```diff
@@ -83,6 +83,34 @@ def blas_libs(self):
...
 
+    @property
+    def scalapack_libs(self):
+        libnames = [ 'libmkl_scalapack' ]
+        if self.spec.satisfies('^openmpi'):
+            libnames.append('libmkl_blacs_openmpi')
+        elif self.spec.satisfies('^mpich@1'):
+            libnames.append('libmkl_blacs')
+        elif self.spec.satisfies('^mpich@2:'):
+            libnames.append('libmkl_blacs_intelmpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/intel-mkl/package.py]: should not that be `libmkl_blacs_mpich`?

Comment (davydden) [inline var/spack/repos/builtin/packages/intel-mkl/package.py]: i recall that i was compiling `trilinos` with `mkl_scalapack_lp64;mkl_blacs_mpich_lp64`, so i would expect `mkl_blacs_mpich` to appear somewhere in this list.

Comment (Zzzoom) [inline var/spack/repos/builtin/packages/intel-mkl/package.py]: I've rechecked the mkl link line advisor for many different versions of MKL and this appears to be accurate. MKL has used intelmpi for MPICH2 (and 3) and Intel MPI for a long time.
Besides, there's no ```libmkl_blacs_mpich_lp64.so``` in my MKL lib directory.

Comment (davydden) [inline var/spack/repos/builtin/packages/intel-mkl/package.py]: I think u r right. My notes were probably incorrect. I also checked link advisor with GCC and OpenMPI, Mpich2 and there is no blacs_mpich. Sorry for confusion.

**Comment by tgamblin** [comment]
@davydden: we haven't used it extensively in the mainline but it IS possible to import from other packages and even to extend them.  These packages uses that functionality when it does:

```python
from spack.pkg.builtin.intel import IntelInstaller
```

Perhaps the code could be moved there in a future PR.  You can also have one package extend another, so if it makes sense, maybe Parallel Studio could extend MKL...

**Comment by tgamblin** [comment]
@Zzzoom: It is valid to do `when='^mpi'`.  But you can also test for `+mpi` if you have a variant and think that is more clear.

**Pull request merged**